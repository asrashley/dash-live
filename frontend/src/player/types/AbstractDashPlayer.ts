import type { OptionMapWithChildren } from "@dashlive/options";
import { KeyParameters } from "./KeyParameters";

export interface DashPlayerProps {
    version?: string;
    autoplay?: boolean;
    videoElement: HTMLVideoElement;
    subtitlesElement?: HTMLDivElement;
    logEvent(eventName: string, text: string): void;
}
export abstract class AbstractDashPlayer {

    // eslint-disable-next-line no-useless-constructor
    constructor(protected props: DashPlayerProps) {
        /* no op */
    }

    abstract initialize(mpd: string, options: OptionMapWithChildren, keys: Map<string, KeyParameters>): Promise<void>;

    abstract destroy(): void;

    protected onCanPlayEvent = async () => {
        const { videoElement } = this.props;
        try {
            await videoElement.play();
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        } catch(err) {
            videoElement.muted = true;
            videoElement.play();
        }
        videoElement.removeEventListener('canplay', this.onCanPlayEvent);
    };

}