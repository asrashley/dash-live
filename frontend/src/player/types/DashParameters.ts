import type { OptionMapWithChildren } from "@dashlive/options";
import type { PeriodJson } from "./PeriodsJson";

export interface DashParameters {
    dash: {
        locationURL: string;
        mediaDuration: string;
        minBufferTime: string;
        mpd_id: string;
        now: string;
        periods: PeriodJson[];
        profiles: string[];
        publishTime: string;
        startNumber: number;
        suggestedPresentationDelay: number;
        timeSource: null;
        title: string;
    },
    options: OptionMapWithChildren;
    url: string;
}
