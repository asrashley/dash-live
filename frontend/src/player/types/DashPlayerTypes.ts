export type DashPlayerTypes = 'native' | 'shaka' | 'dashjs';