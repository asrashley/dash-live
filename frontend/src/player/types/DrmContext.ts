export type DrmContext = {
    laurl: string;
    scheme_id: string;
    version: number;
};
