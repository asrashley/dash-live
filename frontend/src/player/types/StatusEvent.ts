export type StatusEvent = {
    id: number;
    position: number;
    timecode: string;
    event: string;
    text: string;
};