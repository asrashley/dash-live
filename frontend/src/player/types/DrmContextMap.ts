import { DrmContext } from "./DrmContext";

export type DrmContextMap = {
    [drm: string]: DrmContext;
};