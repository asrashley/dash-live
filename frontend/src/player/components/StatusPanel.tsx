import { useComputed, type ReadonlySignal } from "@preact/signals";
import { StatusEvent } from "../types/StatusEvent";
import { createTimeObject, timeObjectToString } from "../utils/formatTimecode";

function StatusRow({ event, timecode, text }: StatusEvent) {
  return (
    <p class="status-row">
      <span className="timecode">{timecode}</span>{event} {text}
    </p>
  );
}

export interface StatusPanelProps {
  events: ReadonlySignal<StatusEvent[]>;
  currentTime: ReadonlySignal<number>;
}

export function StatusPanel({ events, currentTime }: StatusPanelProps) {
  const timecode = useComputed<string>(() =>
    timeObjectToString(createTimeObject(currentTime.value))
  );
  return (
    <div id="status">
      {events.value.map((evt) => (
        <StatusRow key={evt.id} {...evt} />
      ))}
      <div className="play-position">{timecode}</div>
    </div>
  );
}
