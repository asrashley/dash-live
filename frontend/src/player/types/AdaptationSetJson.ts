import { DrmContextMap } from "./DrmContextMap";
import { KeyParameters } from "./KeyParameters";

export type AdaptationSetJson = {
    content_type: string;
    default_kid: string | null;
    drm?: DrmContextMap;
    fileSuffix: string;
    id: number;
    initURL: string;
    keys?: { [kid: string]: KeyParameters };
    lang: null | string;
    mediaURL: string;
    mimeType: string;
    mode: string;
    par: string;
    segmentAlignment: boolean;
    startWithSAP: number;
    timescale: number;
};
