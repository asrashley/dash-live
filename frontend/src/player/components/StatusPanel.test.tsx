import { signal } from "@preact/signals";
import { describe, expect, test } from "vitest";
import { renderWithProviders } from "../../test/renderWithProviders";
import { StatusPanel } from "./StatusPanel";
import { StatusEvent } from "../types/StatusEvent";

describe("StatusPanel", () => {
  const exampleEvents: StatusEvent[] = [
    {
      id: 1,
      position: 0.25,
      timecode: "00:00:00.250",
      event: "loadedmetadata",
      text: "",
    },
    {
      id: 2,
      position: 1,
      timecode: "00:00:01.000",
      event: "canplay",
      text: "",
    },
    {
      id: 3,
      position: 1.5,
      timecode: "00:00:01.500",
      event: "playing",
      text: "",
    },
  ];
  const events = signal<StatusEvent[]>([]);
  const currentTime = signal<number>(0);

  test("matches snapshot", () => {
    currentTime.value = 83;
    events.value = [...exampleEvents];
    const { asFragment, getBySelector } = renderWithProviders(
      <StatusPanel events={events} currentTime={currentTime} />
    );
    const tcElt = getBySelector(".play-position") as HTMLDivElement;
    expect(tcElt?.innerHTML).toEqual("00:01:23");
    expect(asFragment()).toMatchSnapshot();
  });
});
