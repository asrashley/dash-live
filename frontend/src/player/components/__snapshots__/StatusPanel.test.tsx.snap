// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatusPanel > matches snapshot 1`] = `
<DocumentFragment>
  <div
    id="status"
  >
    <p
      class="status-row"
    >
      <span
        class="timecode"
      >
        00:00:00.250
      </span>
      loadedmetadata 
    </p>
    <p
      class="status-row"
    >
      <span
        class="timecode"
      >
        00:00:01.000
      </span>
      canplay 
    </p>
    <p
      class="status-row"
    >
      <span
        class="timecode"
      >
        00:00:01.500
      </span>
      playing 
    </p>
    <div
      class="play-position"
    >
      00:01:23
    </div>
  </div>
</DocumentFragment>
`;
